File: src/MphRead/Mods/CrashReport.cs
using System;
using System.IO;
using System.Text;

namespace MphRead.Mods
{
    /// <summary>
    /// The last thing that runs when nothing else could: what a crash before
    /// the window looks like from the outside, and where it is written down.
    ///
    /// <para>
    /// Not the same thing as <see cref="DebugLog"/>, which is a session-long
    /// transcript kept only when a player has asked for one. This is always
    /// installed and writes exactly one file, because the failure it exists
    /// for is the one nobody can report: the Windows build is a GUI binary
    /// (see <see cref="ConsoleWindow"/>), so an exception on the way to the
    /// first window has nowhere to print. The process disappears, the player
    /// double-clicks the executable again and again, and every symptom there
    /// is is "nothing happens".
    /// </para>
    ///
    /// <para>
    /// So: a file beside the executable, a console with the message in it on
    /// Windows, and a pause when that console belongs to us -- otherwise the
    /// window it was printed into closes with the process.
    /// </para>
    /// </summary>
    public static class CrashReport
    {
        private static bool _installed;
        private static bool _reported;

        /// <summary>Where the last report went, if one was written.</summary>
        public static string? Path { get; private set; }

        /// <summary>
        /// Install the handler. Called first thing in Main, before anything
        /// that can throw -- which on a fresh install is most of startup.
        /// </summary>
        public static void Install()
        {
            if (_installed)
            {
                return;
            }
            _installed = true;
            AppDomain.CurrentDomain.UnhandledException += (_, e) =>
            {
                Report(e.ExceptionObject as Exception, "unhandled");
            };
        }

        /// <summary>
        /// Write one report and, on Windows, make sure it is on the screen.
        ///
        /// Only the first is written: an exception on the way out of another
        /// one would otherwise overwrite the report naming the fault that
        /// actually started it.
        /// </summary>
        public static void Report(Exception? ex, string source)
        {
            if (_reported || ex == null)
            {
                return;
            }
            _reported = true;
            string? path = TryWrite(ex, source);
            Path = path;
            // A GUI binary that got this far has no console at all; without
            // one the message below goes nowhere, which is the whole failure
            // this class exists for.
            if (OperatingSystem.IsWindows())
            {
                ConsoleWindow.Show();
            }
            Console.Error.WriteLine();
            Console.Error.WriteLine($"{Branding.Name} could not start.");
            Console.Error.WriteLine();
            Console.Error.WriteLine($"  {ex.GetType().Name}: {ex.Message}");
            if (path != null)
            {
                Console.Error.WriteLine();
                Console.Error.WriteLine($"The details are in {path}.");
            }
            Console.Error.WriteLine();
            // Started by double-click: the console window is this process's
            // own and closes with it, so everything above would be a flash of
            // black without this.
            if (OperatingSystem.IsWindows() && ConsoleWindow.OwnsItsConsole())
            {
                Console.Error.WriteLine("Press any key to close...");
                try
                {
                    Console.ReadKey(intercept: true);
                }
                catch (Exception)
                {
                    // No key to read is no reason to fail on the way out.
                }
            }
        }

        private static string? TryWrite(Exception ex, string source)
        {
            var text = new StringBuilder();
            text.AppendLine($"{Branding.Name} {Update.BuildVersion.Display}, "
                + $"data format {Program.Version}");
            text.AppendLine($"{DateTime.Now:yyyy-MM-dd HH:mm:ss zzz}");
            text.AppendLine($"{Environment.OSVersion}, .NET {Environment.Version}, "
                + $"64-bit process={Environment.Is64BitProcess}");
            text.AppendLine($"base={AppContext.BaseDirectory}");
            text.AppendLine($"command line={Environment.CommandLine}");
            text.AppendLine($"source={source}");
            text.AppendLine();
            text.AppendLine(ex.ToString());
            // Named .log because LogArchive gathers "*.log", and this has to
            // travel with the rest of them.
            string name = $"{Branding.Name.Replace(" ", "")}-crash-"
                + $"{DateTime.Now:yyyyMMdd-HHmmss}-{Environment.ProcessId}.log";
            // The logs folder first, and not because it is tidy: that is the
            // one directory LogArchive gathers and the share sheet hands out,
            // so a crash written anywhere else is a crash a player on a phone
            // has no way to send. Then beside the executable, where somebody
            // who has just downloaded a desktop release will look, and the
            // temporary directory for an installation under Program Files.
            foreach (string directory in new[]
                     {
                         LogArchive.Directory,
                         Launcher.LauncherPrefs.Directory,
                         AppContext.BaseDirectory,
                         System.IO.Path.GetTempPath()
                     })
            {
                try
                {
                    Directory.CreateDirectory(directory);
                    string path = System.IO.Path.Combine(directory, name);
                    File.WriteAllText(path, text.ToString());
                    return path;
                }
                catch (Exception)
                {
                    // Try the next one; a report that cannot be written is
                    // still a message that can be printed.
                }
            }
            return null;
        }
    }
}

File: src/MphRead.Native/Mods/CrashReport.cpp
#include "CrashReport.hpp"

#include "Branding.hpp"
#include "ConsoleWindow.hpp"
#include "LogShare.hpp"
#include "Launcher/Portable/LauncherPrefs.hpp"
#include "Update/BuildVersion.hpp"
#include "../Program.hpp"
#include "../NativeRuntime/System/AppDomain.hpp"
#include "../NativeRuntime/System/Console.hpp"
#include "../NativeRuntime/System/DateTime.hpp"
#include "../NativeRuntime/System/Encoding.hpp"
#include "../NativeRuntime/System/ExceptionText.hpp"
#include "../NativeRuntime/System/IO.hpp"
#include "../NativeRuntime/System/Runtime.hpp"

#include <array>
#include <exception>
#include <string>

namespace MphRead::Mods
{
    namespace
    {
        namespace Runtime = ::MphRead::NativeRuntime;

        // StringBuilder.AppendLine(value).
        void AppendLine(std::string& text, std::string_view value)
        {
            text.append(value);
            text.append(Runtime::EnvironmentNewLine());
        }

        void AppendLine(std::string& text)
        {
            text.append(Runtime::EnvironmentNewLine());
        }

        // ex.GetType().Name, for an exception held by pointer.
        [[nodiscard]] std::string TypeName(const std::exception_ptr& ex)
        {
            if (!ex)
            {
                return "System.Object";
            }
            try
            {
                std::rethrow_exception(ex);
            }
            catch (const std::exception& value)
            {
                return Runtime::ExceptionTypeName(value);
            }
            catch (...)
            {
                return "System.Object";
            }
        }

        // ex.Message.
        [[nodiscard]] std::string Message(const std::exception_ptr& ex)
        {
            if (!ex)
            {
                return std::string();
            }
            try
            {
                std::rethrow_exception(ex);
            }
            catch (const std::exception& value)
            {
                return value.what();
            }
            catch (...)
            {
                return std::string();
            }
        }
    }

    bool CrashReport::_installed = false;
    bool CrashReport::_reported = false;
    std::optional<std::string> CrashReport::_path{};

    std::optional<std::string> CrashReport::Path()
    {
        return _path;
    }

    void CrashReport::Install()
    {
        if (_installed)
        {
            return;
        }
        _installed = true;
        Runtime::AppDomainAddUnhandledExceptionHandler(
            [](std::exception_ptr ex) { Report(ex, "unhandled"); });
    }

    void CrashReport::Report(const std::exception_ptr& ex, std::string_view source)
    {
        if (_reported || !ex)
        {
            return;
        }
        _reported = true;
        const std::optional<std::string> path = TryWrite(ex, source);
        _path = path;
        // A GUI binary that got this far has no console at all; without one
        // the message below goes nowhere, which is the whole failure this
        // class exists for.
#if defined(_WIN32)
        ConsoleWindow::Show();
#endif
        Runtime::ConsoleErrorWriteLine("");
        Runtime::ConsoleErrorWriteLine(std::string(Branding::Name) + " could not start.");
        Runtime::ConsoleErrorWriteLine("");
        Runtime::ConsoleErrorWriteLine("  " + TypeName(ex) + ": " + Message(ex));
        if (path.has_value())
        {
            Runtime::ConsoleErrorWriteLine("");
            Runtime::ConsoleErrorWriteLine("The details are in " + *path + ".");
        }
        Runtime::ConsoleErrorWriteLine("");
        // Started by double-click: the console window is this process's own
        // and closes with it, so everything above would be a flash of black
        // without this.
#if defined(_WIN32)
        if (ConsoleWindow::OwnsItsConsole())
        {
            Runtime::ConsoleErrorWriteLine("Press any key to close...");
            try
            {
                Runtime::ConsoleReadKeyIntercept();
            }
            catch (const std::exception&)
            {
                // No key to read is no reason to fail on the way out.
            }
        }
#endif
    }

    std::optional<std::string> CrashReport::TryWrite(
        const std::exception_ptr& ex, std::string_view source)
    {
        std::string text;
        AppendLine(text, std::string(Branding::Name) + " " + Update::BuildVersion::Display()
            + ", data format " + Program::Version.ToString());
        AppendLine(text, Runtime::DateTimeToString(
            Runtime::DateTimeNow(), "yyyy-MM-dd HH:mm:ss zzz"));
        AppendLine(text, Runtime::EnvironmentOSVersion() + ", .NET "
            + Runtime::EnvironmentVersion() + ", 64-bit process="
            + (Runtime::EnvironmentIs64BitProcess() ? "True" : "False"));
        AppendLine(text, "base=" + Runtime::AppContextBaseDirectory());
        AppendLine(text, "command line=" + Runtime::EnvironmentCommandLine());
        AppendLine(text, "source=" + std::string(source));
        AppendLine(text);
        AppendLine(text, Runtime::ExceptionToString(ex));
        // Named .log because LogArchive gathers "*.log", and this has to
        // travel with the rest of them.
        std::string name(Branding::Name);
        for (std::size_t space = name.find(' '); space != std::string::npos;
            space = name.find(' '))
        {
            name.erase(space, 1);
        }
        name += "-crash-" + Runtime::DateTimeToString(
            Runtime::DateTimeNow(), "yyyyMMdd-HHmmss")
            + "-" + std::to_string(Runtime::EnvironmentProcessId()) + ".log";
        // The logs folder first, and not because it is tidy: that is the one
        // directory LogArchive gathers and the share sheet hands out, so a
        // crash written anywhere else is a crash a player on a phone has no
        // way to send. Then beside the executable, where somebody who has just
        // downloaded a desktop release will look, and the temporary directory
        // for an installation under Program Files.
        const std::array<std::string, 4> directories = {
            Runtime::Utf16ToUtf8(LogArchive::Directory()),
            Launcher::LauncherPrefs::Directory(),
            Runtime::AppContextBaseDirectory(),
            Runtime::PathGetTempPath()
        };
        for (const std::string& directory : directories)
        {
            try
            {
                Runtime::DirectoryCreateDirectory(directory);
                const std::string path = Runtime::PathCombine(directory, name);
                Runtime::FileWriteAllText(path, text);
                return path;
            }
            catch (const std::exception&)
            {
                // Try the next one; a report that cannot be written is still a
                // message that can be printed.
            }
        }
        return std::nullopt;
    }
}
